version: Reject unexpected values in zulip-git-version.

timabbott opened the pull request:
This allows clients to rely on the Zulip server always presenting a value for its own version that is valid and parseable.

Tested locally with "garbage" or a normal valid value in `zulip-git-version`.

timabbott committed:
version: Reject unexpected values in zulip-git-version.

This allows clients to rely on the Zulip server always presenting a
value for its own version that is valid and parseable.

andersk commented:
We could detect this earlier by removing `--always` from our `git describe` invocation and letting it throw an error.

```diff
-    git describe --always --tags --match='[0-9]*'
+    git describe --tags --match='[0-9]*'
```

I wrote `--always` because I assumed that this isn’t actually a problem—as discussed in the [original thread](https://chat.zulip.org/#narrow/stream/378-api-design/topic/.60zulip_version.60.20in.20server.20settings/near/1728231), one should use `API_FEATURE_LEVEL` for comparisons.

timabbott commented:
We probably still want code like this to fail on unparseable output because the file could be corrupted (by humans most likely). 

Since we do display this value to humans, and have some expectations about its format for display in the gear menu, it seems worth trying to make sure it makes sense, even if code should not be using it for comparisons... there's an argument for replacing the `assert` with just formatting something based on the commit ID we have, but that commit ID is probably useless, so I'm still tempted by this approach, possibly in addition to removing the `--always`?

timabbott committed:
cache-zulip-git-version: Require git describe to succeed.

It's probably better to address and resolve failures in this early
stage than it is to do so later, and I think our code for ensuring the
Git repositories are setup properly in upgrade-zulip-from-git is good
enough at this point that we can rely on it.

timabbott commented:
Added a commit removing `--always`.

timabbott commented:
Hmm removing `--always` makes `./tools/cache-zulip-git-version` fail in our CI environment. Thoughts on how best to resolve that? Those get a partial git history so presumably can't actually do `git describe` properly.

timabbott committed:
tools: Remove zulip-git-version if updating fails.

timabbott commented:
I added an added commit to make `zulip-git-version` get deleted if we get an error trying to update that; I think that is reasonable to avoid it containing invalid output.

alexmv commented on version.py:
```diff
@@ -1,17 +1,29 @@
+import re
 
-ZULIP_VERSION = "9.0-dev+git"
+ZULIP_BASE_VERSION = "9.0-dev+git"
```
`tools/release` tries to parse this out, so will need to be updated for this change.
